File: study_ws/src/py_pkg/py_pkg/activity004_server.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from irobot_interfaces.srv import SetLed
from irobot_interfaces.msg import LEDStates
# Python Specifics..

class LED_Panel_Node(Node): 
    def __init__(self):
        super().__init__("LED_Panel_Node")
        self.get_logger().info('LED Panel Node Server has initiated...')
        self.led_states_publisher_ = self.create_publisher(LEDStates, 'led_states', 10)
        self.server_ = self.create_service(SetLed, 'set_led', self.callback_)

    def callback_(self, request, response):
        try:
            msg = LEDStates()
            msg.led_states = request.input_array
            self.led_states_publisher_.publish(msg)
            response.success = True
            self.get_logger().info(f'LED States Updated: {msg.led_states}')
        except Exception as e:
            response.success = False
            self.get_logger().error(f'LED States Update Failed: {e}')
        return response


def main(args=None):
    rclpy.init(args=args)
    node = LED_Panel_Node() 
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: study_ws/src/cpp_pkg/src/activity004_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "irobot_interfaces/srv/set_led.hpp"
#include "irobot_interfaces/msg/led_states.hpp"

class LED_Panel_Node : public rclcpp::Node 
{
public:
    LED_Panel_Node() : Node("LED_Panel_Node") 
    {
        RCLCPP_INFO(this->get_logger(), "LED Panel Node Server has initiated...");
        using std::placeholders::_1;
        using std::placeholders::_2;

        server_ = this->create_service<irobot_interfaces::srv::SetLed>(
            "set_led", std::bind(&LED_Panel_Node::callback_, this, _1, _2));

        led_states = this->create_publisher<irobot_interfaces::msg::LEDStates>("led_states", 10);
    }

private:
    rclcpp::Service<irobot_interfaces::srv::SetLed>::SharedPtr server_;
    rclcpp::Publisher<irobot_interfaces::msg::LEDStates>::SharedPtr led_states;
    
    void callback_(const std::shared_ptr<irobot_interfaces::srv::SetLed::Request> request,
                   std::shared_ptr<irobot_interfaces::srv::SetLed::Response> response)
    {
        try{
            auto led_state_msg = irobot_interfaces::msg::LEDStates();
            led_state_msg.led_states = request->input_array;
            led_states->publish(led_state_msg);
            std::stringstream ss;
            for (const auto& val : led_state_msg.led_states) ss << val << " ";
            RCLCPP_INFO(this->get_logger(), "LED States Updated: %s", ss.str().c_str());

            response->success = true;
        }
        catch(const std::exception &e){
            RCLCPP_ERROR(this->get_logger(), "Error occurred: %s", e.what());
            response->success = false;
        }
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<LED_Panel_Node>(); 
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
